File: benchmark/src/cupynumeric/benchmarks/nas/mg.py
"""NAS MG using cuPyNumeric views and array expressions.

LIMITATION: The exact NPB sparse right-hand side is generated on the host, as
in NPB-GPU, and uploaded before timing. All residual, restriction,
interpolation, smoothing, and periodic-boundary work stays in cuPyNumeric.
Transfers use separable axis passes and temporary arrays, not JACC's direct
per-cell kernels; array expressions also materialize intermediates. The common
harness times initial zeroing and L2 sum-of-squares, but omits NPB's Linf norm;
see nas/README.md.
"""

import math

import cupynumeric as np
import numpy as host_np

from core import register_benchmark


SEED = 314159265.0
MULTIPLIER = 1220703125.0
EXTREMA = 10
A = (-8.0 / 3.0, 0.0, 1.0 / 6.0, 1.0 / 12.0)
CLASSES = {
    "S": (32, 4, 0.5307707005734e-4),
    "W": (128, 4, 0.6467329375339e-5),
    "A": (256, 4, 0.2433365309069e-5),
    "B": (256, 20, 0.1800564401355e-5),
    "C": (512, 20, 0.5706732285740e-6),
    "D": (1024, 50, 0.1583275060440e-9),
    "E": (2048, 50, 0.8157592357404e-10),
}


def randlc(x, a=MULTIPLIER):
    r23, t23, r46, t46 = 2.0**-23, 2.0**23, 2.0**-46, 2.0**46
    a1 = math.trunc(r23 * a)
    a2 = a - t23 * a1
    x1 = math.trunc(r23 * x)
    x2 = x - t23 * x1
    t1 = a1 * x2 + a2 * x1
    z = t1 - t23 * math.trunc(r23 * t1)
    t3 = t23 * z + a2 * x2
    nxt = t3 - t46 * math.trunc(r46 * t3)
    return nxt, r46 * nxt


def insert_extreme(values, indices, value, index, largest):
    if (largest and value <= values[0]) or (not largest and value >= values[0]):
        return
    values[0], indices[0] = value, index
    for i in range(len(values) - 1):
        ordered = values[i] <= values[i + 1] if largest else values[i] >= values[i + 1]
        if ordered:
            break
        values[i], values[i + 1] = values[i + 1], values[i]
        indices[i], indices[i + 1] = indices[i + 1], indices[i]


def rhs_host(n):
    lows, highs = [1.0] * EXTREMA, [0.0] * EXTREMA
    low_indices, high_indices = [(0, 0, 0)] * EXTREMA, [(0, 0, 0)] * EXTREMA
    seed = SEED
    for k in range(1, n + 1):
        for j in range(1, n + 1):
            for i in range(1, n + 1):
                seed, value = randlc(seed)
                index = (i, j, k)
                insert_extreme(lows, low_indices, value, index, False)
                insert_extreme(highs, high_indices, value, index, True)
    rhs = host_np.zeros((n + 2, n + 2, n + 2), dtype=host_np.float64)
    for index in low_indices:
        rhs[index] = -1.0
    for index in high_indices:
        rhs[index] = 1.0
    comm3(rhs)
    return rhs


def comm3(u):
    u[0, 1:-1, 1:-1] = u[-2, 1:-1, 1:-1]
    u[-1, 1:-1, 1:-1] = u[1, 1:-1, 1:-1]
    u[:, 0, 1:-1] = u[:, -2, 1:-1]
    u[:, -1, 1:-1] = u[:, 1, 1:-1]
    u[:, :, 0] = u[:, :, -2]
    u[:, :, -1] = u[:, :, 1]
    return u


def resid(r, u, v):
    c = (slice(1, -1),) * 3
    xm, xp = slice(0, -2), slice(2, None)
    m, p = slice(0, -2), slice(2, None)
    mid = slice(1, -1)
    r[c] = (
        v[c]
        - A[0] * u[c]
        - A[2]
        * (
            u[mid, m, m]
            + u[mid, p, m]
            + u[mid, m, p]
            + u[mid, p, p]
            + u[xm, mid, m]
            + u[xp, mid, m]
            + u[xm, mid, p]
            + u[xp, mid, p]
            + u[xm, m, mid]
            + u[xp, m, mid]
            + u[xm, p, mid]
            + u[xp, p, mid]
        )
        - A[3]
        * (
            u[xm, m, m]
            + u[xp, m, m]
            + u[xm, p, m]
            + u[xp, p, m]
            + u[xm, m, p]
            + u[xp, m, p]
            + u[xm, p, p]
            + u[xp, p, p]
        )
    )
    return comm3(r)


def psinv(u, r, coeff):
    c = (slice(1, -1),) * 3
    xm, xp, mid = slice(0, -2), slice(2, None), slice(1, -1)
    m, p = slice(0, -2), slice(2, None)
    u[c] += (
        coeff[0] * r[c]
        + coeff[1]
        * (
            r[xm, mid, mid]
            + r[xp, mid, mid]
            + r[mid, m, mid]
            + r[mid, p, mid]
            + r[mid, mid, m]
            + r[mid, mid, p]
        )
        + coeff[2]
        * (
            r[mid, m, m]
            + r[mid, p, m]
            + r[mid, m, p]
            + r[mid, p, p]
            + r[xm, mid, m]
            + r[xp, mid, m]
            + r[xm, mid, p]
            + r[xp, mid, p]
            + r[xm, m, mid]
            + r[xp, m, mid]
            + r[xm, p, mid]
            + r[xp, p, mid]
        )
    )
    return comm3(u)


def axis_last(array, axis):
    if axis == 2:
        return array, (0, 1, 2)
    permutation = (1, 2, 0) if axis == 0 else (0, 2, 1)
    inverse = tuple(permutation.index(i) for i in range(3))
    return np.transpose(array, permutation), inverse


def restrict_axis(array, axis):
    back, inverse = axis_last(array, axis)
    d1, d2, n = back.shape
    physical = n - 2
    # Legate cannot reshape a sliced store. Materialize two contiguous slabs,
    # pair neighboring cells in the last dimension, then reduce each pair.
    left = back[:, :, 1 : n - 1].copy()
    right = back[:, :, 2:n].copy()
    paired_shape = (d1, d2, physical // 2, 2)
    reduced = left.reshape(paired_shape).sum(axis=3) + right.reshape(paired_shape).sum(
        axis=3
    )
    return reduced if axis == 2 else np.transpose(reduced, inverse)


def restrict(coarse, fine):
    reduced = restrict_axis(fine, 0)
    reduced = restrict_axis(reduced, 1)
    reduced = restrict_axis(reduced, 2)
    coarse[1:-1, 1:-1, 1:-1] = reduced / 16.0
    return comm3(coarse)


def interp_axis(array, axis, weights):
    back, inverse = axis_last(array, axis)
    d1, d2, n = back.shape
    lo = back[:, :, : n - 1].copy().reshape(d1, d2, n - 1, 1)
    hi = back[:, :, 1:n].copy().reshape(d1, d2, n - 1, 1)
    interpolated = (lo + weights * (hi - lo)).reshape(d1, d2, 2 * (n - 1))
    result = interpolated if axis == 2 else np.transpose(interpolated, inverse)
    return result.copy()


def interp(fine, coarse, weights):
    interpolated = interp_axis(coarse, 0, weights)
    interpolated = interp_axis(interpolated, 1, weights)
    interpolated = interp_axis(interpolated, 2, weights)
    fine += interpolated


def norm2(residual):
    interior = residual[1:-1, 1:-1, 1:-1]
    return np.sum(interior * interior)


class NASMultiGrid:
    name = "nas_mg"

    def __init__(self, T, N, M, **kwargs):
        self.T, self.N, self.M = T, N, M
        self.class_name = str(kwargs.pop("class", "S")).upper()
        if kwargs:
            raise ValueError(f"Unknown NAS MG options: {', '.join(kwargs)}")
        if self.class_name not in CLASSES:
            raise ValueError(f"Unknown NAS MG class {self.class_name}")
        n = CLASSES[self.class_name][0]
        if T is not np.float64 or (N, M) != (n, n):
            raise ValueError(
                f"NAS MG class {self.class_name} requires Float64, N=M={n}"
            )

    def dims(self):
        return self.N, self.M

    def initialize(self):
        n = CLASSES[self.class_name][0]
        sizes = [2**level + 2 for level in range(1, int(math.log2(n)) + 1)]
        return {
            "u": [np.zeros((s, s, s), dtype=np.float64) for s in sizes],
            "r": [np.zeros((s, s, s), dtype=np.float64) for s in sizes],
            "rhs": np.asarray(rhs_host(n)),
            "weights": np.asarray([0.0, 0.5]).reshape(1, 1, 1, 2),
        }

    def run(self, state):
        niter = CLASSES[self.class_name][1]
        coeff = (
            (-3.0 / 8.0, 1.0 / 32.0, -1.0 / 64.0, 0.0)
            if self.class_name in ("S", "W", "A")
            else (-3.0 / 17.0, 1.0 / 33.0, -1.0 / 61.0, 0.0)
        )
        u, r, rhs, weights = state["u"], state["r"], state["rhs"], state["weights"]
        for level in u:
            level.fill(0.0)
        resid(r[-1], u[-1], rhs)
        norm2(r[-1])
        for _ in range(niter):
            for level in range(len(u) - 1, 0, -1):
                restrict(r[level - 1], r[level])
            u[0].fill(0.0)
            psinv(u[0], r[0], coeff)
            for level in range(1, len(u) - 1):
                u[level].fill(0.0)
                interp(u[level], u[level - 1], weights)
                resid(r[level], u[level], r[level])
                psinv(u[level], r[level], coeff)
            interp(u[-1], u[-2], weights)
            resid(r[-1], u[-1], rhs)
            psinv(u[-1], r[-1], coeff)
            resid(r[-1], u[-1], rhs)
        return norm2(r[-1])

    def correctness_dims(self):
        return self.N, self.M

    def check_correctness(self):
        n, _, reference = CLASSES[self.class_name]
        squared = self.run(self.initialize())
        norm = math.sqrt(float(host_np.asarray(squared)) / n**3)
        return "pass" if abs((norm - reference) / reference) <= 1.0e-8 else "fail"


register_benchmark("nas_mg", NASMultiGrid)

File: benchmark/src/cunumeric/benchmarks/nas/mg.jl
# LIMITATION: NPB MG's exact sparse right-hand side is generated on the host,
# as in NPB-GPU, then uploaded once before timing. The V-cycle itself uses
# cuNumeric views and broadcasts, allowing Legate to partition every level.
# Restriction/interpolation are separable axis passes with temporary arrays,
# not JACC's direct per-cell kernels. Views are no-ops; explicit slice/store
# operations express the regions. The common harness times initial zeroing
# and L2 sum-of-squares, but omits NPB's Linf norm; see nas/README.md.

mutable struct CuNumericNASMGState{U,R,V,W}
    u::U
    r::R
    rhs::V
    interp_weights::W
end

const CuNumericMGIndex = Union{Colon,AbstractUnitRange{<:Integer}}

function cunumeric_mg_slice(index::Colon, n)
    return (0, n)
end
cunumeric_mg_slice(index::AbstractUnitRange, n) = (Int(first(index)) - 1, Int(last(index)))

# cuNumeric's core currently defines contiguous slicing through rank 2. MG is
# rank 3, so provide the identical view operation locally until core generalizes
# those methods. The returned NDArray shares its parent's Legate store.
function Base.getindex(
    array::cuNumeric.NDArray{T,3}, i::CuNumericMGIndex,
    j::CuNumericMGIndex, k::CuNumericMGIndex,
) where {T}
    @boundscheck checkbounds(array, i, j, k)
    slices = cuNumeric.slice_array(
        cunumeric_mg_slice(i, size(array, 1)),
        cunumeric_mg_slice(j, size(array, 2)),
        cunumeric_mg_slice(k, size(array, 3)),
    )
    return cuNumeric.nda_get_slice(array, slices)
end

function Base.setindex!(
    lhs::cuNumeric.NDArray{T,3}, rhs::cuNumeric.NDArray,
    i::CuNumericMGIndex, j::CuNumericMGIndex, k::CuNumericMGIndex,
) where {T}
    view = lhs[i, j, k]
    copyto!(view, rhs)
    cuNumeric.destroy!(view)
    return rhs
end

function initialize(b::NASMultiGrid{Float64}; mod=cuNumeric)
    p = validate_nas_mg(b)
    sizes = nas_mg_level_sizes(p)
    u = [mod.zeros(Float64, n, n, n) for n in sizes]
    r = [mod.zeros(Float64, n, n, n) for n in sizes]
    rhs = mod.NDArray(nas_mg_rhs(p))
    weights = mod.reshape(mod.NDArray([0.0, 0.5]), 1, 1, 1, 2)
    return (CuNumericNASMGState(u, r, rhs, weights),)
end

function cunumeric_mg_back(array, axis)
    axis == 3 && return array, (1, 2, 3)
    permutation = axis == 1 ? (2, 3, 1) : (1, 3, 2)
    return permutedims(array, permutation), invperm(permutation)
end

function cunumeric_mg_restrict_axis(array, axis)
    back, inverse = cunumeric_mg_back(array, axis)
    d1, d2, n = size(back)
    physical = n - 2
    # cuNumeric reshapes stores in C order, so move the active dimension to
    # the back before pairing neighboring points. Legate cannot reshape a
    # sliced store; materialize only these two contiguous shifted slabs.
    left = copy(back[:, :, 2:(n - 1)])
    right = copy(back[:, :, 3:n])
    paired_shape = (d1, d2, physical ÷ 2, 2)
    reduced = cuNumeric.reshape(
        sum(cuNumeric.reshape(left, paired_shape); dims=4) .+
        sum(cuNumeric.reshape(right, paired_shape); dims=4),
        d1, d2, physical ÷ 2,
    )
    return axis == 3 ? reduced : permutedims(reduced, inverse)
end

function cunumeric_mg_restrict!(coarse, fine)
    reduced = cunumeric_mg_restrict_axis(fine, 1)
    reduced = cunumeric_mg_restrict_axis(reduced, 2)
    reduced = cunumeric_mg_restrict_axis(reduced, 3)
    n = size(coarse, 1)
    interior = coarse[2:(n - 1), 2:(n - 1), 2:(n - 1)]
    interior .= reduced ./ 16.0
    cuNumeric.destroy!(interior)
    return nas_mg_comm3!(coarse)
end

function cunumeric_mg_interp_axis(array, axis, weights)
    back, inverse = cunumeric_mg_back(array, axis)
    d1, d2, n = size(back)
    lo = copy(back[:, :, 1:(n - 1)])
    hi = copy(back[:, :, 2:n])
    lo4 = cuNumeric.reshape(lo, d1, d2, n - 1, 1)
    hi4 = cuNumeric.reshape(hi, d1, d2, n - 1, 1)
    mixed = lo4 .+ weights .* (hi4 .- lo4)
    interpolated = cuNumeric.reshape(mixed, d1, d2, 2(n - 1))
    result = axis == 3 ? interpolated : permutedims(interpolated, inverse)
    # The next interpolation axis permutes this result. Give it an independent
    # store instead of composing a transpose with the reshape view above.
    return copy(result)
end

function cunumeric_mg_interp!(fine, coarse, weights)
    interpolated = cunumeric_mg_interp_axis(coarse, 1, weights)
    interpolated = cunumeric_mg_interp_axis(interpolated, 2, weights)
    interpolated = cunumeric_mg_interp_axis(interpolated, 3, weights)
    fine .+= interpolated
    return fine
end

function cunumeric_mg_cycle!(s, c)
    finest = length(s.u)
    for level in finest:-1:2
        cunumeric_mg_restrict!(s.r[level - 1], s.r[level])
    end
    fill!(s.u[1], 0.0)
    nas_mg_psinv!(s.u[1], s.r[1], c)
    for level in 2:(finest - 1)
        fill!(s.u[level], 0.0)
        cunumeric_mg_interp!(s.u[level], s.u[level - 1], s.interp_weights)
        nas_mg_resid!(s.r[level], s.u[level], s.r[level])
        nas_mg_psinv!(s.u[level], s.r[level], c)
    end
    cunumeric_mg_interp!(s.u[end], s.u[end - 1], s.interp_weights)
    nas_mg_resid!(s.r[end], s.u[end], s.rhs)
    nas_mg_psinv!(s.u[end], s.r[end], c)
    return nothing
end

function cunumeric_mg_norm2(residual)
    n = size(residual, 1)
    interior = residual[2:(n - 1), 2:(n - 1), 2:(n - 1)]
    squared = sum(interior .* interior)
    cuNumeric.destroy!(interior)
    return squared
end

function run!(b::NASMultiGrid, s::CuNumericNASMGState)
    p = nas_mg_parameters(b.class)
    c = nas_mg_smoother(b.class)
    foreach(x -> fill!(x, 0.0), s.u)
    nas_mg_resid!(s.r[end], s.u[end], s.rhs)
    cunumeric_mg_norm2(s.r[end])
    for _ in 1:p.niter
        cunumeric_mg_cycle!(s, c)
        nas_mg_resid!(s.r[end], s.u[end], s.rhs)
    end
    return cunumeric_mg_norm2(s.r[end])
end

function check_benchmark_correctness(b::NASMultiGrid, gs::GlobalSettings; mod=cuNumeric)
    state = only(initialize(b; mod))
    squared = run!(b, state)
    norm = sqrt(cuNumeric.@allowscalar squared[] / Float64(b.N)^3)
    return nas_mg_verified(b.class, norm) ? "pass" : "fail"
end
